lib/src/metta/interpreter.rs

```diff
@@ -1047,20 +1048,25 @@ fn type_cast(space: Atom, atom: Atom, expected_type: Atom, bindings: Bindings) -
         once((return_atom(atom), bindings))
     } else {
         let space = space.as_gnd::<DynSpace>().unwrap();
-        let first_match = get_atom_types(space, &atom).into_iter()
-            .map(|actual_type| match_types(&expected_type, &actual_type, bindings.clone()))
-            .filter(|res| res.is_ok())
-            .flat_map(|res| {
-                match res {
+        let types = get_atom_types(space, &atom);
+        let actual_types = types.clone().into_iter()
+            .filter(AtomType::is_valid)
+            .map(AtomType::into_atom);
+        let mut idx: i64 = 0;
+        let mut first_match: MatchResultIter;
+        for actual_type in actual_types {
+            idx += 1;
+            let matched = match_types(&expected_type, &actual_type, bindings.clone());
+            if matched.is_ok() {
+                first_match = match matched {
                     Ok(it) => it,
-                    Err(_) => panic!("Unexpected state"),
-                }
-            })
-            .next();
-        match first_match {
-            Some(bindings) => once((return_atom(atom), bindings)),
-            None => once((return_atom(error_atom(atom, BAD_TYPE_SYMBOL)), bindings)),
+                    Err(_) => unreachable!(),
+                };
+                return once((return_atom(atom), first_match.next().unwrap()));
+            }
         }
+        let actual_type = types.iter().nth((idx-1) as usize).unwrap();
+        once((return_atom(Atom::expr([ERROR_SYMBOL, atom.clone(), Atom::expr([BAD_TYPE_SYMBOL, Atom::gnd(Number::Integer(idx)), expected_type.clone(), actual_type.as_atom().clone()])])), bindings))
```

Comment:
Two issues here:
1. `types` vector is unnecessary cloned above
2. `idx` is not position inside the atom but returned as such inside `BadType` error

I tried to fix them in https://github.com/DaddyWesker/hyperon-experimental/pull/29 but there is still question what to return in such situation see my another comment.